clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -245,12 +247,23 @@ public int partitionLeaderEpoch() {
             protected Record readNext(long baseOffset, long baseTimestamp, int baseSequence, Long logAppendTime) {
                 try {
                     return DefaultRecord.readFrom(inputStream, baseOffset, baseTimestamp, baseSequence, logAppendTime);
+                } catch (EOFException e) {
+                    throw new InvalidRecordException("Incorrect declared batch size, premature EOF reached");
                 } catch (IOException e) {
                     throw new KafkaException("Failed to decompress record stream", e);
                 }
             }
 
             @Override
+            protected boolean hasRemaining() {
+                try {
+                    return inputStream.read() == -1;
```

Comment:
The name is wrong. This should be `hasNoneRemaining()` or, even better (your suggestion) `ensureNoneRemaining()`. Previously it was called `eofReached()` but @ijuma reqested it to be changed to `hasRemaining()`, and I didn't check the condition :)